tests/sles4sap/qesapdeployment/deploy.pm

```diff
@@ -10,37 +10,22 @@ use Mojo::Base 'publiccloud::basetest';
 use testapi;
 use qesapdeployment;
 
-sub wait_for_ssh {
-    my ($host) = @_;
-    my $timeout //= bmwqemu::scale_timeout(600);
-    my $start_time = time();
-    my $check_port = 1;
-
-    # Looping until reaching timeout or passing two conditions :
-    # - SSH port 22 is reachable
-    # - journalctl got message about reaching one of certain targets
-    while ((my $duration = time() - $start_time) < $timeout) {
-        if ($check_port) {
-            $check_port = 0 if (script_run('nc -vz -w 1 ' . $host . ' 22', quiet => 1) == 0);
-        }
-        else {
-            return $duration;
-        }
-        sleep 5;
-    }
-    die 'Timed out while waiting for ssh to be available in the CSP instances';
-}
-
 sub run {
     my $ret = qesap_execute(cmd => 'terraform', verbose => 1, timeout => 1800);
     die "'qesap.py terraform' return: $ret" if ($ret);
     my $inventory = qesap_get_inventory(get_required_var('PUBLIC_CLOUD_PROVIDER'));
     upload_logs($inventory, failok => 1);
     my @remote_ips = qesap_remote_hana_public_ips;
     record_info 'Remote IPs', join(' - ', @remote_ips);
-    foreach my $host (@remote_ips) { wait_for_ssh $host; }
+    foreach my $host (@remote_ips) {
+        die 'Timed out while waiting for ssh to be available in the CSP instances' if qesap_wait_for_ssh(host => $host) == -1;
+    }
     $ret = qesap_execute(cmd => 'ansible', cmd_options => '--profile', verbose => 1, timeout => 3600);
-    die "'qesap.py ansible' return: $ret" if ($ret);
+    if ($ret)
+    {
+        qesap_cluster_logs();
+        die "'qesap.py ansible' return: $ret";
+    }
```

Comment:
Why not keep the line as `die "'qesap.py ansible' return: $ret" if ($ret);` and add the call to `qesap_cluster_logs()` in the `post_fail_hook`?

Reply:
generally speaking I'm trying a new idea here. It is about keep all the error handling of the expected errors in the `run` part. I like to leave only the handling of the  `unexpected` events in the `post_fail_hook` ... like an error during the log collection. I can survive if logs are not collected at a failure, but I like to try to organize the code to avoid that anything could prevent the deployment destroy to be executed. Does it make sense?

Reply:
Yes, it does. But keep in mind that the `die` there is going to call the `post_fail_hook` ... could we end up with a test module that it's uploading the same set of logs twice? Wouldn't that be confusing?